networkit/cpp/centrality/KadabraBetweenness.cpp

```diff
@@ -0,0 +1,526 @@
+
+	for (i = 0; i < status->k; ++i) {
+		bet[i] = status->approxTop[i] / (double)status->nPairs;
+	}
+
+	if (absolute) {
+		for (i = 0; i < status->k; ++i) {
+			errL[i] = err;
+			errU[i] = err;
+		}
+	} else {
+		errU[0] = std::max(err, (bet[0] - bet[1]) / 2.);
+		errL[0] = 10;
+		for (i = 1; i < k; ++i) {
+			errL[i] = std::max(err, (bet[i - 1] - bet[i]) / 2.);
+			errU[i] = std::max(err, (bet[i] - bet[i + 1]) / 2.);
+		}
+		for (i = k; i < status->k; ++i) {
+			errL[i] = 10;
+			errU[i] = std::max(err, bet[k - 1] + (bet[k - 1] - bet[k]) / 2. - bet[i]);
+		}
+		for (i = 0; i < k - 1; ++i) {
+			if (bet[i] - bet[i + 1] < maxErr) {
+				errL[i] = err;
+				errU[i] = err;
+				errL[i + 1] = err;
+				errU[i + 1] = err;
+			}
+		}
+		for (i = k + 1; i < status->k; ++i) {
+			if (bet[k] - bet[i] < maxErr) {
+				errL[k] = err;
+				errU[k] = err;
+				errL[i] = err;
+				errU[i] = err;
+			}
+		}
+	}
+}
+
+void KadabraBetweenness::computeDeltaGuess() {
+	double a = 0,
+	       b = 1. / err / err * std::log(n * 4 * (1 - balancingFactor) / delta),
+	       c = (a + b) / 2;
+	double sum;
+
+	Status status(unionSample);
+	getStatus(&status, true);
+
+	std::vector<double> bet(status.k);
+	std::vector<double> errL(status.k);
+	std::vector<double> errU(status.k);
+
+	computeBetErr(&status, bet, errL, errU);
+
+	for (count i = 0; i < unionSample; ++i) {
+		count v = status.top[i];
+		approxSum[v] = approxSum[v] / (double)nPairs;
+	}
+
+	while (b - a > err / 10.) {
+		c = (b + a) / 2.;
+		sum = 0;
+		for (count i = 0; i < unionSample; ++i) {
+			sum += std::exp(-c * errL[i] * errL[i] / bet[i]);
+			sum += std::exp(-c * errU[i] * errU[i] / bet[i]);
+		}
+
+		sum += std::exp(-c * errL[unionSample - 1] * errL[unionSample - 1] /
+		                bet[unionSample - 1]) *
+		       (n - unionSample);
+		sum += std::exp(-c * errU[unionSample - 1] * errU[unionSample - 1] /
+		                bet[unionSample - 1]) *
+		       (n - unionSample);
+
+		if (sum >= delta / 2. * (1 - balancingFactor)) {
+			a = c;
+		} else {
+			b = c;
+		}
+	}
+
+	deltaLMinGuess = std::exp(-b * errL[unionSample - 1] * errL[unionSample - 1] /
+	                          bet[unionSample - 1]) +
+	                 delta * balancingFactor / 4. / (double)n;
+	deltaUMinGuess = std::exp(-b * errU[unionSample - 1] * errU[unionSample - 1] /
+	                          bet[unionSample - 1]) +
+	                 delta * balancingFactor / 4. / (double)n;
+
+	std::fill(deltaLGuess.begin(), deltaLGuess.end(), deltaLMinGuess);
+	std::fill(deltaUGuess.begin(), deltaUGuess.end(), deltaUMinGuess);
+
+	for (count i = 0; i < unionSample; ++i) {
+		node v = status.top[i];
+		deltaLGuess[v] = std::exp(-b * errL[i] * errL[i] / bet[i]) +
+		                 delta * balancingFactor / 4. / (double)n;
+		deltaUGuess[v] = std::exp(-b * errU[i] * errU[i] / bet[i]) +
+		                 delta * balancingFactor / 4. / (double)n;
+	}
+}
+
+void KadabraBetweenness::computeApproxParallel(const bool normalize) {
```

Comment:
We sometimes need a sequential version of this method (e.g. in line 343), right? Why don't we add a parameter to control if it parallel as in `getStatus()`?

Reply:
We use this function in line 315 and 358, both calls are executed in a sequential section of the code. I think we can keep it as it is.

Reply:
I see. OK, let's keep it that way, then.